# # Error Handling
### <p style="color:Tomato">Learned the basics of error handling while finding common names for U.S. Congress members.<p/>
some basic exploration and manipulation with legislators.csv, and laid the groundwork for our "names project".

In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'

#### <p style="color:Gray">Missing Data<p/><hr>
working with **legislators.csv** <br/>
:which records information on every historical member of the U.S. Conress.

some rows contain misiing values for some columns.<br/>

> we'll explore some of the errors that occur when we ignore missing values and how to handle them

When exploring data, it's often useful to extract the unique elements in a list.<br/>
데이터를 탐색 할 때 고유 한 요소를 목록에서 추출하는 것이 유용한 경우가 많습니다. 

#### <p style="color:Gray">Set<p/><hr>
If you add an element to a set that it already contains, the set will ignore it. <br/>
The items in a set are unordered, while each item in a list has an index.<br/>
Sets don't have indexes, the items in a set may display in a different order each time you print it.<br/>
<br/>
이미 포함 된 집합에 요소를 추가하면 집합에서이를 무시합니다.<br/>
집합의 항목은 정렬되지 않은 반면, 목록의 각 항목에는 색인이 있습니다. Sets에는 인덱스가 없으므로 세트를 인쇄 할 때마다 세트의 항목이 다른 순서로 표시 될 수 있습니다.

In [2]:
["Dog", "Cat", "Hippo", "Dog", "Cat", "Dog", "Dog", "Cat"]
unique_animals = set(["Dog", "Cat", "Hippo", "Dog", "Cat", "Dog", "Dog", "Cat"])
print(unique_animals)

['Dog', 'Cat', 'Hippo', 'Dog', 'Cat', 'Dog', 'Dog', 'Cat']

{'Hippo', 'Dog', 'Cat'}


In [3]:
unique_animals.add('Tiger')
unique_animals.remove('Dog')
print(unique_animals)
list(unique_animals)

{'Hippo', 'Tiger', 'Cat'}


['Hippo', 'Tiger', 'Cat']

In [4]:
import csv
f = open('legislators.csv', 'r')
csvreader = csv.reader(f)
legi_data = list(csvreader)
legi_data[:10]

[['last_name', 'first_name', 'birthday', 'gender', 'type', 'state', 'party'],
 ['Bassett', 'Richard', '1745-04-02', 'M', 'sen', 'DE', 'Anti-Administration'],
 ['Bland', 'Theodorick', '1742-03-21', '', 'rep', 'VA', ''],
 ['Burke', 'Aedanus', '1743-06-16', '', 'rep', 'SC', ''],
 ['Carroll', 'Daniel', '1730-07-22', 'M', 'rep', 'MD', ''],
 ['Clymer', 'George', '1739-03-16', 'M', 'rep', 'PA', ''],
 ['Contee', 'Benjamin', '', 'M', 'rep', 'MD', ''],
 ['Dalton', 'Tristram', '1738-05-28', '', 'sen', 'MA', 'Pro-Administration'],
 ['Elmer', 'Jonathan', '1745-11-29', 'M', 'sen', 'NJ', 'Pro-Administration'],
 ['Few', 'William', '1748-06-08', 'M', 'sen', 'GA', 'Anti-Administration']]

In [5]:
class Dataset:
    def __init__(self, data):
        self.header = data[0]
        self.data = data[1:]
    def column(self, label):
        if label not in self.header:
            return None
        index = 0
        for idx, element in enumerate(self.header):
            if label == element:
                index = idx
        column = []
        for row in self.data:
            column.append(row[index])
        return column
    def count_unique(self, label):
        unique_results = set(self.column(label))
        count = len(unique_results)
        return count
    def set_column(self, label):
        unique_results = set(self.column(label))
        return unique_results

In [6]:
legi_dataset = Dataset(legi_data)
header = legi_dataset.header
gender = legi_dataset.column('gender')
print(header)
print(gender[:10])

['last_name', 'first_name', 'birthday', 'gender', 'type', 'state', 'party']
['M', '', '', 'M', 'M', 'M', '', 'M', 'M', 'M']


In [7]:
gender = set(gender)
print(gender)

{'', 'M', 'F'}


In [8]:
gender_def = legi_dataset.set_column('gender')
print(gender_def)

{'', 'M', 'F'}


In [9]:
gender = []
for item in legi_data[1:]:
    gender.append(item[3])
    
gender = set(gender)
print(gender)

{'', 'M', 'F'}


<p style="color:Tomato">I need to remove null values<p/>
<p style="color:Gray">Print out datas and inspect the data<p/><hr>
working with **legislators.csv** <br/>
:which records information on every historical member of the U.S. Conress.

some rows contain misiing values for some columns.<br/>

> we'll explore some of the errors that occur when we ignore missing values and how to handle them
When you have a fresh dataset, it's always a good idea to look for any patterns.
* Missing data
    * some files contain empty fields. Others may use a string like N/A to indicate missing values.
* Values that don't make intuitive sense
    * A legislator with a birhday in 2050, for example, would indicate a problem with the data.
* Recurring themes 
    * One theme in this dataset is that the overwhelming majority of legislators are male.

<br/>

새로운 데이터 세트를 가지고있을 때는 항상 다음과 같은 패턴을 찾아 보는 것이 좋습니다.
* 누락된 데이터
    * 일부 파일에는 빈 필드가 있습니다. 다른 사람들은 N / A와 같은 문자열을 사용하여 누락 된 값을 나타낼 수 있습니다.
* 직관적으로 이해할 수 없는 가치
    * 예를 들어, 2050 년 생일을 가진 입법자는 데이터에 문제가 있음을 나타냅니다.
* 되풀이되는 테마
    * 이 데이터 세트의 하나의 주제는 압도적으로 다수의 의원이 남성이라는 점입니다. 

In [10]:
header = legi_dataset.header
print(header)

['last_name', 'first_name', 'birthday', 'gender', 'type', 'state', 'party']


In [11]:
legislators = legi_dataset.data
print(legislators[:10])

[['Bassett', 'Richard', '1745-04-02', 'M', 'sen', 'DE', 'Anti-Administration'], ['Bland', 'Theodorick', '1742-03-21', '', 'rep', 'VA', ''], ['Burke', 'Aedanus', '1743-06-16', '', 'rep', 'SC', ''], ['Carroll', 'Daniel', '1730-07-22', 'M', 'rep', 'MD', ''], ['Clymer', 'George', '1739-03-16', 'M', 'rep', 'PA', ''], ['Contee', 'Benjamin', '', 'M', 'rep', 'MD', ''], ['Dalton', 'Tristram', '1738-05-28', '', 'sen', 'MA', 'Pro-Administration'], ['Elmer', 'Jonathan', '1745-11-29', 'M', 'sen', 'NJ', 'Pro-Administration'], ['Few', 'William', '1748-06-08', 'M', 'sen', 'GA', 'Anti-Administration'], ['Floyd', 'William', '1734-12-17', 'M', 'rep', 'NY', '']]


In [12]:
# Print out party and inspect the values. And print out legislators and inspect the data.
party = []
for item in legislators:
    party.append(item[6])
party = set(party)    
print(party)
print(legislators[:10])


{'', 'Liberal Republican', 'Anti-Jacksonian', 'Anti-Administration', 'Jackson Republican', 'Democratic Republican', 'Readjuster Democrat', 'Progressive', 'Prohibitionist', 'Democrat-Liberal', 'Jacksonian', 'Silver Republican', 'Free Silver', 'New Progressive', 'Socialist', 'Democrat', 'Progressive Republican', 'Ind. Republican-Democrat', 'Ind. Republican', 'Constitutional Unionist', 'Conservative Republican', 'Nonpartisan', 'Independent Democrat', 'American Labor', 'Republican', 'Law and Order', 'Whig', 'Unconditional Unionist', 'Union', 'Anti Jackson', 'Ind. Democrat', 'Federalist', 'Adams', 'Conservative', 'Unionist', 'Independent', 'Unknown', 'Adams Democrat', 'Farmer-Labor', 'Nullifier', 'Anti Jacksonian', 'Readjuster', 'Popular Democrat', 'National Greenbacker', 'Anti-Lecompton Democrat', 'Populist', 'Union Labor', 'Union Democrat', 'Democratic and Union Labor', 'Crawford Republican', 'American', 'Pro-Administration', 'Free Soil', 'States Rights', 'Ind. Whig', 'Jackson', 'Coalitio

> gender and party columns contained an **empty string('')**. This indicates that one or more of the rows in the data have missing values in those columns. <br/>
Missing values are very common in real world data analysis, since the people compiling the datasets often don't have full information.<br/>
<br/>
성별 및 파티 열의 집합 표현에 빈 문자열 ( '')이 포함되어 있음을 알았을 수 있습니다. 이는 데이터의 하나 이상의 행에 해당 열에 누락 값이 있음을 나타냅니다. 누락 된 값은 실제 데이터 분석에서 매우 일반적입니다. 데이터 집합을 컴파일하는 사람들이 완전한 정보를 가지고 있지 않기 때문입니다.

<p style="color:Gray">Strategies to address missing data<p/><hr>
you can use one of the following strategies to address missing data.
* Remove any rows that contain missing data
* populate the empty fields with a specified value.
* populate the empty fields with a calculated value.
* Use analysis techniques that work with missing data

In [13]:
# focus on populating empty fields with a specified value
rows = [ ["Bassett", "Richard", "1745-04-02", "M", "sen", "DE", "Anti-Administration"],
    ["Bland", "Theodorick", "1742-03-21", "", "rep", "VA", ""]]
for row in rows:
    if row[6] == "":
        row[6] = "No Party"

In [14]:
print(header)

['last_name', 'first_name', 'birthday', 'gender', 'type', 'state', 'party']


In [15]:
# Replace any missing values in the gender column of legislators with the string "M"
for row in legislators:
    if row[3] == "":
        row[3] = "M"

In [16]:
gender = []
for item in legislators:
    gender.append(item[3])

In [17]:
print(set(gender))

{'M', 'F'}


In [18]:
print(legislators[:30])

[['Bassett', 'Richard', '1745-04-02', 'M', 'sen', 'DE', 'Anti-Administration'], ['Bland', 'Theodorick', '1742-03-21', 'M', 'rep', 'VA', ''], ['Burke', 'Aedanus', '1743-06-16', 'M', 'rep', 'SC', ''], ['Carroll', 'Daniel', '1730-07-22', 'M', 'rep', 'MD', ''], ['Clymer', 'George', '1739-03-16', 'M', 'rep', 'PA', ''], ['Contee', 'Benjamin', '', 'M', 'rep', 'MD', ''], ['Dalton', 'Tristram', '1738-05-28', 'M', 'sen', 'MA', 'Pro-Administration'], ['Elmer', 'Jonathan', '1745-11-29', 'M', 'sen', 'NJ', 'Pro-Administration'], ['Few', 'William', '1748-06-08', 'M', 'sen', 'GA', 'Anti-Administration'], ['Floyd', 'William', '1734-12-17', 'M', 'rep', 'NY', ''], ['Gale', 'George', '1756-06-03', 'M', 'rep', 'MD', ''], ['Grayson', 'William', '', 'M', 'sen', 'VA', 'Anti-Administration'], ['Grout', 'Jonathan', '1737-07-23', 'M', 'rep', 'MA', ''], ['Hawkins', 'Benjamin', '1754-08-15', 'M', 'sen', 'NC', 'Pro-Administration'], ['Huntington', 'Benjamin', '1736-04-19', 'M', 'rep', 'CT', ''], ['Izard', 'Ralph', 

<p style="color:Gray">Identify historical naming trends, and explore how popular names have changed form 1820 to today<p/><hr>
* looking for the most common names of U.S. legislators
* The year of birth

> The birthday column has the format 1820-01-02, which is hard to work with.<br/>
1. It's common to reformat values to simplify them.<br/>
2. split the data into its component parts

>생일 형식이 1820-01-02이면 작업하기 어렵습니다. 
1. 따라서 간단하게 값을 다시 형식화해야 합니다.
2. 날짜를 구성요소로 나눕니다.

In [19]:
print(header)

['last_name', 'first_name', 'birthday', 'gender', 'type', 'state', 'party']


In [20]:
date = "1820-01-02"
parts = date.split("-")
print(parts)

['1820', '01', '02']


In [21]:
birth_years = []
for item in legislators:
    parts = item[2].split('-')
    # print(parts[0])
    birth_years.append(parts[0])
print(birth_years[:10])

['1745', '1742', '1743', '1730', '1739', '', '1738', '1745', '1748', '1734']


* Converting a column to a different data type is a common operation in data analysis.

<p style="color:Gray">string form -> integers<p/><hr>
* date data
* To find the average year in which legislators born
* convert the data to integers
* int() function
> we just extracted the year on the previous screen, but it's in string form. To find the average year in which legislators were born, we'll need to convert the data to integers first. We can perform this conversion with the int() function. 
    1. handle the missing value - an empty string can't be converted to an integer.
    2. 

#### <p style="color:Tomato">Try/except<p/>
<p style="color:Gray">We can Manage errors with something known as a try/except block<p/><hr>

> Not all errors should halt execution, thogh. Sometimes we expect **a certain type of errors**, and want to handle it in a specific way that allows the code to complete.<br/>
If you surround the code that causes an error with a try/except block, the error will be handled, and the code will continue to run.

> 모든 오류가 실행을 중단해야하는 것은 아닙니다. 때로는 특정 유형의 오류가 예상되며 코드를 완료 할 수있는 특정 방식으로 처리하려고합니다.  try / except 블록에서 오류를 일으키는 코드를 둘러싸면 오류가 처리되고 코드가 계속 실행됩니다.

In [22]:
try:
    int('1')
except Exception:
    print("There was an error")

1

In [23]:
try:
    int('I am not a int')
except Exception:
    print("There was an error")

There was an error


In [24]:
string = float('hello')

ValueError: could not convert string to float: 'hello'

In [25]:
try:
    float("hello")
except Exception:
    print("Error converting to float.")

Error converting to float.


> try/eception
1. The python interpreter will try to run int(''), and generate a ValueError.
2. Instead of stopping the code from executing, it will be handled by the except statement, which will print the message "There was an error"
3. Python interpreter will countinue to run any lines of code that come after the cxcpt statment. 

> try/eception
1. 파이썬 인터프리터는 int()를 실행하고 ValueError를 나타내려고 한다.
2. 코드는 계속 실행되고, except문에 의해 처리된다. 오류 발생의 메세지를 출력한다.  
3. except문 다음에 나오는 코드를 계속 실행한다. 

<p style="color:Gray">Exception class<p/><hr>
When the Python interpreter generates an exception, it actually creates an instance of the Exception class.
* Has certain properties that help us **debug the error**
* Access the instance of the Exception class in the except statement body.<br/>
<br/>

파이썬 인터프리터가 예외를 생성 할 때 실제로 Exception 클래스의 인스턴스를 생성합니다.
* 오류를 디버그하는데 유용한 특정 속성을 가지고 있습니다. 
* except문의 본문에서 Exception 클래스의 인스턴스에 접근할 수 있습니다. 

In [26]:
try:
    int('')
except Exception as exc:
    print(type(exc)) # we can then access the vaiable in the except statement body.

<class 'ValueError'>


In [27]:
try:
    int('')
except Exception as exc:
    print(str(exc))

invalid literal for int() with base 10: ''


In [28]:
try:
    int('')
except Exception as exc:
    print(type(exc)) # Print the type of the Exception instance.
    print(str(exc)) # Convert the Exception instance to a string and print it out.

<class 'ValueError'>
invalid literal for int() with base 10: ''


In [29]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
for i in numbers:
    try:
        int('')
    except Exception:
        print("There was an error")

There was an error
There was an error
There was an error
There was an error
There was an error
There was an error
There was an error
There was an error
There was an error
There was an error


<p style="color:Gray">Pass<p/><hr>
However, there are times when we don't want to do anything specific to handle eroors; we just want the code to keep running. <br/>
While the pass keyword doesn't actually do anything, it's a valid statement body. <br/>
<br/>
그러나 오류를 처리하기 위해 특정 작업을 수행하고 싶지 않은 경우가 있습니다. 코드를 계속 실행하기 만하면됩니다. pass 키워드는 실제로 아무 것도하지 않지만 유효한 문 본문입니다. 

In [30]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
for i in numbers:
    try:
        int('')
    except Exception:
        pass

In [31]:
converted_years = []
for year in birth_years:
    try:
        year = int(year)
    except Exception:
        pass
    converted_years.append(year)

<p style="color:Gray">Convert all of the birth years in legislators to integers<p/><hr>
* To change the items in a list of lists.
* Need to loop over the top-level list(items)
* Convert all of the values in the second column of items to integers.
<br/>
<br/>
* 목록안의 목록의 항목을 변경합니다.
* 최상위 목록이 필요
* 항목의 두 번째 열에있는 모든 값을 정수로 변환

In [32]:
items = [          
        [1, "1", 2],    
        [2, "", 3],
        [5, "5", 3]
        ]

for item in items:
    item[1] = int(item[1])

ValueError: invalid literal for int() with base 10: ''

In [33]:
legi_dataset.header
checking_birthday = legi_dataset.column('birthday') # year-month-day

['last_name', 'first_name', 'birthday', 'gender', 'type', 'state', 'party']

In [34]:
for row in legislators:
    birthday = row[2]
    birth_year = birthday.split("-")[0] # birth_year = row[2].split("-")[0]
    try:
        birth_year = int(birth_year)
    except Exception:
        birth_year = 0
    row.append(birth_year)

In [35]:
print(legislators[:10])

[['Bassett', 'Richard', '1745-04-02', 'M', 'sen', 'DE', 'Anti-Administration', 1745], ['Bland', 'Theodorick', '1742-03-21', 'M', 'rep', 'VA', '', 1742], ['Burke', 'Aedanus', '1743-06-16', 'M', 'rep', 'SC', '', 1743], ['Carroll', 'Daniel', '1730-07-22', 'M', 'rep', 'MD', '', 1730], ['Clymer', 'George', '1739-03-16', 'M', 'rep', 'PA', '', 1739], ['Contee', 'Benjamin', '', 'M', 'rep', 'MD', '', 0], ['Dalton', 'Tristram', '1738-05-28', 'M', 'sen', 'MA', 'Pro-Administration', 1738], ['Elmer', 'Jonathan', '1745-11-29', 'M', 'sen', 'NJ', 'Pro-Administration', 1745], ['Few', 'William', '1748-06-08', 'M', 'sen', 'GA', 'Anti-Administration', 1748], ['Floyd', 'William', '1734-12-17', 'M', 'rep', 'NY', '', 1734]]


<p style="color:Gray">Using Chronological order situation<p/><hr>
>While exploring the dataset, you may have noticed that the legislators appear in roughly chronological order. 
<br/>

* The values generally **appear in chronological order**, we can loop through each year and replace any 0 values with the values from **the previous rows**
* By doing this, we'll make sure each legislator without a birth year is assigned one that's relatively close to the actual date.
<br/>
<br/>
* 값은 일반적으로 시간순을 표시되기 때문에 각 연도마다 loop를 돌리고 0값을 이전 행의 값으로 바꿀 수 있습니다. 
* 이렇게 함으로써 출생년도가 없는 의원에게 실제 날짜와 비교적 가깝게 값을 대입할 수 있습니다. 

1. We finished parsing the birth years to integers
2. we have several birth years with the value 0 
3. we replaced missing values with a fixed value M.
4. Using Chronological order Situation, to assign the value. 

<br/>

* 우리는 출생 년도를 정수로 파싱을 마쳤습니다.
* 0의 값을 몇 개 가지고 있습니다. 
* 누락된 값을 고정 값 M으로 바꾸었습니다. 
* 시간순으로 나타나는 데이터의 상황을 이용해서 0값에 적절한 값들을 대입시킬 수 있습니다. 

### <p style="color:Tomato">★<p/>

In [40]:
last_value = 1
for row in legislators:
    # print(row[7])
    # print("last_value:")
    # print(last_value)
    if row[7] == 0:
        row[7] = last_value
    last_value = row[7]

1745
last_value:
1
1742
last_value:
1745
1743
last_value:
1742
1730
last_value:
1743
1739
last_value:
1730
1739
last_value:
1739
1738
last_value:
1739
1745
last_value:
1738
1748
last_value:
1745
1734
last_value:
1748
1756
last_value:
1734
1756
last_value:
1756
1737
last_value:
1756
1754
last_value:
1737
1736
last_value:
1754
1736
last_value:
1736
1727
last_value:
1736
1733
last_value:
1727
1732
last_value:
1733
1737
last_value:
1732
1739
last_value:
1737
1734
last_value:
1739
1740
last_value:
1734
1745
last_value:
1740
1728
last_value:
1745
1728
last_value:
1728
1738
last_value:
1728
1737
last_value:
1738
1739
last_value:
1737
1744
last_value:
1739
1744
last_value:
1744
1761
last_value:
1744
1756
last_value:
1761
1752
last_value:
1756
1737
last_value:
1752
1745
last_value:
1737
1744
last_value:
1745
1742
last_value:
1744
1726
last_value:
1742
1726
last_value:
1726
1733
last_value:
1726
1741
last_value:
1733
1741
last_value:
1741
1756
last_value:
1741
1756
last_value:
1756
1721
last_val

1756
last_value:
1758
1778
last_value:
1756
1776
last_value:
1778
1765
last_value:
1776
1767
last_value:
1765
1778
last_value:
1767
1768
last_value:
1778
1784
last_value:
1768
1777
last_value:
1784
1777
last_value:
1777
1770
last_value:
1777
1761
last_value:
1770
1774
last_value:
1761
1768
last_value:
1774
1777
last_value:
1768
1772
last_value:
1777
1766
last_value:
1772
1782
last_value:
1766
1770
last_value:
1782
1772
last_value:
1770
1763
last_value:
1772
1774
last_value:
1763
1770
last_value:
1774
1770
last_value:
1770
1770
last_value:
1770
1761
last_value:
1770
1759
last_value:
1761
1763
last_value:
1759
1765
last_value:
1763
1816
last_value:
1765
1774
last_value:
1816
1785
last_value:
1774
1757
last_value:
1785
1756
last_value:
1757
1756
last_value:
1756
1781
last_value:
1756
1764
last_value:
1781
1764
last_value:
1764
1745
last_value:
1764
1774
last_value:
1745
1773
last_value:
1774
1776
last_value:
1773
1775
last_value:
1776
1745
last_value:
1775
1745
last_value:
1745
1745
last_

1763
1788
last_value:
1779
1788
last_value:
1788
1776
last_value:
1788
1772
last_value:
1776
1776
last_value:
1772
1776
last_value:
1776
1773
last_value:
1776
1782
last_value:
1773
1792
last_value:
1782
1780
last_value:
1792
1790
last_value:
1780
1780
last_value:
1790
1764
last_value:
1780
1777
last_value:
1764
1777
last_value:
1777
1777
last_value:
1777
1772
last_value:
1777
1772
last_value:
1772
1772
last_value:
1772
1797
last_value:
1772
1779
last_value:
1797
1791
last_value:
1779
1785
last_value:
1791
1786
last_value:
1785
1764
last_value:
1786
1793
last_value:
1764
1793
last_value:
1793
1797
last_value:
1793
1792
last_value:
1797
1784
last_value:
1792
1786
last_value:
1784
1798
last_value:
1786
1789
last_value:
1798
1763
last_value:
1789
1779
last_value:
1763
1785
last_value:
1779
1792
last_value:
1785
1792
last_value:
1792
1793
last_value:
1792
1793
last_value:
1793
1788
last_value:
1793
1790
last_value:
1788
1790
last_value:
1790
1788
last_value:
1790
1769
last_value:
1788
1783


1778
last_value:
1783
1778
last_value:
1778
1780
last_value:
1778
1789
last_value:
1780
1789
last_value:
1789
1780
last_value:
1789
1793
last_value:
1780
1789
last_value:
1793
1799
last_value:
1789
1789
last_value:
1799
1786
last_value:
1789
1795
last_value:
1786
1795
last_value:
1795
1795
last_value:
1795
1795
last_value:
1795
1805
last_value:
1795
1779
last_value:
1805
1776
last_value:
1779
1794
last_value:
1776
1794
last_value:
1794
1789
last_value:
1794
1800
last_value:
1789
1783
last_value:
1800
1803
last_value:
1783
1794
last_value:
1803
1785
last_value:
1794
1766
last_value:
1785
1794
last_value:
1766
1789
last_value:
1794
1798
last_value:
1789
1790
last_value:
1798
1786
last_value:
1790
1775
last_value:
1786
1799
last_value:
1775
1782
last_value:
1799
1801
last_value:
1782
1801
last_value:
1801
1796
last_value:
1801
1796
last_value:
1796
1796
last_value:
1796
1792
last_value:
1796
1806
last_value:
1792
1780
last_value:
1806
1783
last_value:
1780
1794
last_value:
1783
1783
last_

last_value:
1806
1790
last_value:
1797
1800
last_value:
1790
1807
last_value:
1800
1787
last_value:
1807
1795
last_value:
1787
1785
last_value:
1795
1809
last_value:
1785
1798
last_value:
1809
1798
last_value:
1798
1802
last_value:
1798
1803
last_value:
1802
1805
last_value:
1803
1808
last_value:
1805
1808
last_value:
1808
1801
last_value:
1808
1804
last_value:
1801
1787
last_value:
1804
1804
last_value:
1787
1790
last_value:
1804
1799
last_value:
1790
1800
last_value:
1799
1798
last_value:
1800
1792
last_value:
1798
1792
last_value:
1792
1795
last_value:
1792
1788
last_value:
1795
1808
last_value:
1788
1808
last_value:
1808
1808
last_value:
1808
1793
last_value:
1808
1787
last_value:
1793
1797
last_value:
1787
1794
last_value:
1797
1795
last_value:
1794
1802
last_value:
1795
1802
last_value:
1802
1786
last_value:
1802
1786
last_value:
1786
1808
last_value:
1786
1799
last_value:
1808
1807
last_value:
1799
1810
last_value:
1807
1783
last_value:
1810
1801
last_value:
1783
1793
last_value

1797
1815
last_value:
1814
1800
last_value:
1815
1811
last_value:
1800
1808
last_value:
1811
1808
last_value:
1808
1786
last_value:
1808
1808
last_value:
1786
1817
last_value:
1808
1807
last_value:
1817
1822
last_value:
1807
1799
last_value:
1822
1806
last_value:
1799
1811
last_value:
1806
1810
last_value:
1811
1801
last_value:
1810
1805
last_value:
1801
1802
last_value:
1805
1794
last_value:
1802
1816
last_value:
1794
1801
last_value:
1816
1808
last_value:
1801
1808
last_value:
1808
1812
last_value:
1808
1782
last_value:
1812
1799
last_value:
1782
1800
last_value:
1799
1799
last_value:
1800
1804
last_value:
1799
1805
last_value:
1804
1803
last_value:
1805
1802
last_value:
1803
1803
last_value:
1802
1810
last_value:
1803
1811
last_value:
1810
1797
last_value:
1811
1782
last_value:
1797
1799
last_value:
1782
1793
last_value:
1799
1818
last_value:
1793
1793
last_value:
1818
1808
last_value:
1793
1808
last_value:
1808
1806
last_value:
1808
1791
last_value:
1806
1791
last_value:
1791
1811


1813
1806
last_value:
1813
1806
last_value:
1806
1812
last_value:
1806
1798
last_value:
1812
1802
last_value:
1798
1805
last_value:
1802
1824
last_value:
1805
1819
last_value:
1824
1805
last_value:
1819
1806
last_value:
1805
1816
last_value:
1806
1815
last_value:
1816
1804
last_value:
1815
1807
last_value:
1804
1824
last_value:
1807
1794
last_value:
1824
1822
last_value:
1794
1805
last_value:
1822
1826
last_value:
1805
1811
last_value:
1826
1818
last_value:
1811
1811
last_value:
1818
1811
last_value:
1811
1809
last_value:
1811
1798
last_value:
1809
1802
last_value:
1798
1811
last_value:
1802
1808
last_value:
1811
1812
last_value:
1808
1808
last_value:
1812
1803
last_value:
1808
1812
last_value:
1803
1817
last_value:
1812
1815
last_value:
1817
1805
last_value:
1815
1813
last_value:
1805
1819
last_value:
1813
1819
last_value:
1819
1823
last_value:
1819
1795
last_value:
1823
1802
last_value:
1795
1822
last_value:
1802
1808
last_value:
1822
1820
last_value:
1808
1809
last_value:
1820
1800


last_value:
1826
1817
last_value:
1812
1829
last_value:
1817
1814
last_value:
1829
1812
last_value:
1814
1820
last_value:
1812
1818
last_value:
1820
1816
last_value:
1818
1810
last_value:
1816
1828
last_value:
1810
1818
last_value:
1828
1820
last_value:
1818
1807
last_value:
1820
1822
last_value:
1807
1826
last_value:
1822
1818
last_value:
1826
1830
last_value:
1818
1825
last_value:
1830
1819
last_value:
1825
1821
last_value:
1819
1819
last_value:
1821
1835
last_value:
1819
1814
last_value:
1835
1815
last_value:
1814
1828
last_value:
1815
1829
last_value:
1828
1815
last_value:
1829
1821
last_value:
1815
1811
last_value:
1821
1823
last_value:
1811
1822
last_value:
1823
1812
last_value:
1822
1815
last_value:
1812
1816
last_value:
1815
1813
last_value:
1816
1820
last_value:
1813
1818
last_value:
1820
1826
last_value:
1818
1832
last_value:
1826
1828
last_value:
1832
1824
last_value:
1828
1809
last_value:
1824
1827
last_value:
1809
1816
last_value:
1827
1799
last_value:
1816
1820
last_value

1828
last_value:
1823
1822
last_value:
1828
1819
last_value:
1822
1820
last_value:
1819
1820
last_value:
1820
1826
last_value:
1820
1838
last_value:
1826
1818
last_value:
1838
1822
last_value:
1818
1829
last_value:
1822
1837
last_value:
1829
1831
last_value:
1837
1829
last_value:
1831
1814
last_value:
1829
1835
last_value:
1814
1816
last_value:
1835
1822
last_value:
1816
1808
last_value:
1822
1828
last_value:
1808
1840
last_value:
1828
1816
last_value:
1840
1838
last_value:
1816
1843
last_value:
1838
1817
last_value:
1843
1822
last_value:
1817
1828
last_value:
1822
1828
last_value:
1828
1811
last_value:
1828
1821
last_value:
1811
1820
last_value:
1821
1816
last_value:
1820
1815
last_value:
1816
1818
last_value:
1815
1823
last_value:
1818
1827
last_value:
1823
1823
last_value:
1827
1845
last_value:
1823
1827
last_value:
1845
1821
last_value:
1827
1826
last_value:
1821
1814
last_value:
1826
1816
last_value:
1814
1822
last_value:
1816
1815
last_value:
1822
1819
last_value:
1815
1837
last_

1850
1838
last_value:
1834
1832
last_value:
1838
1814
last_value:
1832
1812
last_value:
1814
1835
last_value:
1812
1814
last_value:
1835
1819
last_value:
1814
1823
last_value:
1819
1833
last_value:
1823
1815
last_value:
1833
1831
last_value:
1815
1828
last_value:
1831
1830
last_value:
1828
1831
last_value:
1830
1817
last_value:
1831
1827
last_value:
1817
1824
last_value:
1827
1813
last_value:
1824
1814
last_value:
1813
1818
last_value:
1814
1840
last_value:
1818
1834
last_value:
1840
1837
last_value:
1834
1827
last_value:
1837
1828
last_value:
1827
1829
last_value:
1828
1819
last_value:
1829
1831
last_value:
1819
1831
last_value:
1831
1828
last_value:
1831
1825
last_value:
1828
1836
last_value:
1825
1832
last_value:
1836
1839
last_value:
1832
1842
last_value:
1839
1821
last_value:
1842
1829
last_value:
1821
1829
last_value:
1829
1840
last_value:
1829
1823
last_value:
1840
1826
last_value:
1823
1833
last_value:
1826
1833
last_value:
1833
1830
last_value:
1833
1825
last_value:
1830
1840


1844
last_value:
1846
1824
last_value:
1844
1852
last_value:
1824
1817
last_value:
1852
1840
last_value:
1817
1845
last_value:
1840
1829
last_value:
1845
1833
last_value:
1829
1820
last_value:
1833
1824
last_value:
1820
1835
last_value:
1824
1834
last_value:
1835
1852
last_value:
1834
1842
last_value:
1852
1843
last_value:
1842
1849
last_value:
1843
1821
last_value:
1849
1825
last_value:
1821
1835
last_value:
1825
1828
last_value:
1835
1840
last_value:
1828
1839
last_value:
1840
1840
last_value:
1839
1832
last_value:
1840
1846
last_value:
1832
1844
last_value:
1846
1842
last_value:
1844
1840
last_value:
1842
1838
last_value:
1840
1840
last_value:
1838
1843
last_value:
1840
1838
last_value:
1843
1837
last_value:
1838
1848
last_value:
1837
1841
last_value:
1848
1847
last_value:
1841
1850
last_value:
1847
1839
last_value:
1850
1841
last_value:
1839
1831
last_value:
1841
1836
last_value:
1831
1848
last_value:
1836
1834
last_value:
1848
1830
last_value:
1834
1830
last_value:
1830
1831
last_

last_value:
1854
1838
last_value:
1847
1836
last_value:
1838
1835
last_value:
1836
1854
last_value:
1835
1839
last_value:
1854
1826
last_value:
1839
1854
last_value:
1826
1849
last_value:
1854
1850
last_value:
1849
1831
last_value:
1850
1835
last_value:
1831
1840
last_value:
1835
1841
last_value:
1840
1837
last_value:
1841
1856
last_value:
1837
1847
last_value:
1856
1843
last_value:
1847
1843
last_value:
1843
1846
last_value:
1843
1833
last_value:
1846
1849
last_value:
1833
1841
last_value:
1849
1840
last_value:
1841
1846
last_value:
1840
1824
last_value:
1846
1853
last_value:
1824
1864
last_value:
1853
1856
last_value:
1864
1837
last_value:
1856
1825
last_value:
1837
1860
last_value:
1825
1839
last_value:
1860
1852
last_value:
1839
1832
last_value:
1852
1854
last_value:
1832
1863
last_value:
1854
1846
last_value:
1863
1835
last_value:
1846
1846
last_value:
1835
1842
last_value:
1846
1837
last_value:
1842
1821
last_value:
1837
1862
last_value:
1821
1844
last_value:
1862
1835
last_value

1852
last_value:
1842
1852
last_value:
1852
1856
last_value:
1852
1835
last_value:
1856
1850
last_value:
1835
1843
last_value:
1850
1848
last_value:
1843
1839
last_value:
1848
1844
last_value:
1839
1858
last_value:
1844
1826
last_value:
1858
1858
last_value:
1826
1848
last_value:
1858
1857
last_value:
1848
1856
last_value:
1857
1829
last_value:
1856
1827
last_value:
1829
1862
last_value:
1827
1850
last_value:
1862
1847
last_value:
1850
1867
last_value:
1847
1848
last_value:
1867
1843
last_value:
1848
1851
last_value:
1843
1850
last_value:
1851
1836
last_value:
1850
1852
last_value:
1836
1861
last_value:
1852
1860
last_value:
1861
1843
last_value:
1860
1860
last_value:
1843
1844
last_value:
1860
1841
last_value:
1844
1854
last_value:
1841
1841
last_value:
1854
1846
last_value:
1841
1851
last_value:
1846
1854
last_value:
1851
1858
last_value:
1854
1841
last_value:
1858
1841
last_value:
1841
1838
last_value:
1841
1839
last_value:
1838
1840
last_value:
1839
1848
last_value:
1840
1843
last_

last_value:
1841
1854
last_value:
1869
1853
last_value:
1854
1861
last_value:
1853
1873
last_value:
1861
1842
last_value:
1873
1862
last_value:
1842
1867
last_value:
1862
1875
last_value:
1867
1877
last_value:
1875
1858
last_value:
1877
1873
last_value:
1858
1873
last_value:
1873
1867
last_value:
1873
1872
last_value:
1867
1844
last_value:
1872
1870
last_value:
1844
1859
last_value:
1870
1852
last_value:
1859
1864
last_value:
1852
1861
last_value:
1864
1850
last_value:
1861
1878
last_value:
1850
1855
last_value:
1878
1853
last_value:
1855
1849
last_value:
1853
1875
last_value:
1849
1858
last_value:
1875
1846
last_value:
1858
1852
last_value:
1846
1864
last_value:
1852
1845
last_value:
1864
1873
last_value:
1845
1850
last_value:
1873
1858
last_value:
1850
1859
last_value:
1858
1853
last_value:
1859
1833
last_value:
1853
1854
last_value:
1833
1844
last_value:
1854
1858
last_value:
1844
1861
last_value:
1858
1864
last_value:
1861
1859
last_value:
1864
1852
last_value:
1859
1860
last_value

1879
1879
last_value:
1848
1867
last_value:
1879
1882
last_value:
1867
1849
last_value:
1882
1869
last_value:
1849
1870
last_value:
1869
1857
last_value:
1870
1882
last_value:
1857
1855
last_value:
1882
1867
last_value:
1855
1856
last_value:
1867
1854
last_value:
1856
1857
last_value:
1854
1852
last_value:
1857
1858
last_value:
1852
1863
last_value:
1858
1861
last_value:
1863
1870
last_value:
1861
1870
last_value:
1870
1854
last_value:
1870
1867
last_value:
1854
1861
last_value:
1867
1860
last_value:
1861
1871
last_value:
1860
1869
last_value:
1871
1856
last_value:
1869
1864
last_value:
1856
1865
last_value:
1864
1850
last_value:
1865
1859
last_value:
1850
1840
last_value:
1859
1870
last_value:
1840
1870
last_value:
1870
1873
last_value:
1870
1861
last_value:
1873
1855
last_value:
1861
1868
last_value:
1855
1858
last_value:
1868
1875
last_value:
1858
1864
last_value:
1875
1870
last_value:
1864
1853
last_value:
1870
1860
last_value:
1853
1881
last_value:
1860
1872
last_value:
1881
1863


1876
1868
last_value:
1880
1882
last_value:
1868
1867
last_value:
1882
1835
last_value:
1867
1864
last_value:
1835
1853
last_value:
1864
1863
last_value:
1853
1869
last_value:
1863
1871
last_value:
1869
1855
last_value:
1871
1856
last_value:
1855
1879
last_value:
1856
1870
last_value:
1879
1881
last_value:
1870
1877
last_value:
1881
1856
last_value:
1877
1870
last_value:
1856
1862
last_value:
1870
1868
last_value:
1862
1878
last_value:
1868
1870
last_value:
1878
1861
last_value:
1870
1877
last_value:
1861
1890
last_value:
1877
1878
last_value:
1890
1870
last_value:
1878
1868
last_value:
1870
1865
last_value:
1868
1868
last_value:
1865
1866
last_value:
1868
1849
last_value:
1866
1863
last_value:
1849
1873
last_value:
1863
1863
last_value:
1873
1858
last_value:
1863
1878
last_value:
1858
1869
last_value:
1878
1849
last_value:
1869
1869
last_value:
1849
1851
last_value:
1869
1866
last_value:
1851
1883
last_value:
1866
1869
last_value:
1883
1879
last_value:
1869
1877
last_value:
1879
1858


1877
1879
last_value:
1870
1884
last_value:
1879
1864
last_value:
1884
1870
last_value:
1864
1876
last_value:
1870
1866
last_value:
1876
1854
last_value:
1866
1853
last_value:
1854
1867
last_value:
1853
1875
last_value:
1867
1880
last_value:
1875
1859
last_value:
1880
1865
last_value:
1859
1853
last_value:
1865
1878
last_value:
1853
1890
last_value:
1878
1857
last_value:
1890
1871
last_value:
1857
1875
last_value:
1871
1883
last_value:
1875
1861
last_value:
1883
1866
last_value:
1861
1874
last_value:
1866
1869
last_value:
1874
1860
last_value:
1869
1882
last_value:
1860
1870
last_value:
1882
1868
last_value:
1870
1893
last_value:
1868
1877
last_value:
1893
1883
last_value:
1877
1872
last_value:
1883
1884
last_value:
1872
1878
last_value:
1884
1860
last_value:
1878
1873
last_value:
1860
1874
last_value:
1873
1867
last_value:
1874
1878
last_value:
1867
1892
last_value:
1878
1867
last_value:
1892
1861
last_value:
1867
1880
last_value:
1861
1886
last_value:
1880
1886
last_value:
1886
1884


1888
1892
last_value:
1899
1894
last_value:
1892
1881
last_value:
1894
1877
last_value:
1881
1895
last_value:
1877
1880
last_value:
1895
1899
last_value:
1880
1866
last_value:
1899
1887
last_value:
1866
1865
last_value:
1887
1888
last_value:
1865
1903
last_value:
1888
1895
last_value:
1903
1892
last_value:
1895
1898
last_value:
1892
1886
last_value:
1898
1874
last_value:
1886
1890
last_value:
1874
1889
last_value:
1890
1880
last_value:
1889
1895
last_value:
1880
1887
last_value:
1895
1870
last_value:
1887
1897
last_value:
1870
1905
last_value:
1897
1900
last_value:
1905
1892
last_value:
1900
1889
last_value:
1892
1888
last_value:
1889
1892
last_value:
1888
1907
last_value:
1892
1908
last_value:
1907
1890
last_value:
1908
1887
last_value:
1890
1898
last_value:
1887
1898
last_value:
1898
1888
last_value:
1898
1903
last_value:
1888
1893
last_value:
1903
1880
last_value:
1893
1879
last_value:
1880
1888
last_value:
1879
1870
last_value:
1888
1880
last_value:
1870
1903
last_value:
1880
1885


1918
last_value:
1909
1898
last_value:
1918
1887
last_value:
1898
1903
last_value:
1887
1892
last_value:
1903
1906
last_value:
1892
1880
last_value:
1906
1884
last_value:
1880
1902
last_value:
1884
1887
last_value:
1902
1918
last_value:
1887
1893
last_value:
1918
1886
last_value:
1893
1878
last_value:
1886
1885
last_value:
1878
1910
last_value:
1885
1895
last_value:
1910
1903
last_value:
1895
1894
last_value:
1903
1889
last_value:
1894
1902
last_value:
1889
1904
last_value:
1902
1899
last_value:
1904
1880
last_value:
1899
1890
last_value:
1880
1886
last_value:
1890
1876
last_value:
1886
1884
last_value:
1876
1879
last_value:
1884
1920
last_value:
1879
1875
last_value:
1920
1883
last_value:
1875
1891
last_value:
1883
1894
last_value:
1891
1892
last_value:
1894
1907
last_value:
1892
1910
last_value:
1907
1875
last_value:
1910
1883
last_value:
1875
1903
last_value:
1883
1885
last_value:
1903
1921
last_value:
1885
1884
last_value:
1921
1895
last_value:
1884
1899
last_value:
1895
1895
last_

1874
1920
last_value:
1880
1898
last_value:
1920
1919
last_value:
1898
1901
last_value:
1919
1906
last_value:
1901
1917
last_value:
1906
1903
last_value:
1917
1912
last_value:
1903
1897
last_value:
1912
1916
last_value:
1897
1918
last_value:
1916
1885
last_value:
1918
1911
last_value:
1885
1902
last_value:
1911
1901
last_value:
1902
1915
last_value:
1901
1899
last_value:
1915
1912
last_value:
1899
1904
last_value:
1912
1917
last_value:
1904
1923
last_value:
1917
1920
last_value:
1923
1921
last_value:
1920
1890
last_value:
1921
1890
last_value:
1890
1879
last_value:
1890
1893
last_value:
1879
1897
last_value:
1893
1901
last_value:
1897
1897
last_value:
1901
1918
last_value:
1897
1896
last_value:
1918
1887
last_value:
1896
1925
last_value:
1887
1913
last_value:
1925
1895
last_value:
1913
1905
last_value:
1895
1896
last_value:
1905
1892
last_value:
1896
1893
last_value:
1892
1922
last_value:
1893
1903
last_value:
1922
1897
last_value:
1903
1910
last_value:
1897
1887
last_value:
1910
1929


1925
1905
last_value:
1938
1914
last_value:
1905
1916
last_value:
1914
1918
last_value:
1916
1923
last_value:
1918
1912
last_value:
1923
1901
last_value:
1912
1916
last_value:
1901
1915
last_value:
1916
1904
last_value:
1915
1903
last_value:
1904
1916
last_value:
1903
1932
last_value:
1916
1912
last_value:
1932
1896
last_value:
1912
1919
last_value:
1896
1910
last_value:
1919
1923
last_value:
1910
1912
last_value:
1923
1899
last_value:
1912
1919
last_value:
1899
1916
last_value:
1919
1934
last_value:
1916
1914
last_value:
1934
1934
last_value:
1914
1912
last_value:
1934
1907
last_value:
1912
1926
last_value:
1907
1922
last_value:
1926
1916
last_value:
1922
1924
last_value:
1916
1928
last_value:
1924
1903
last_value:
1928
1915
last_value:
1903
1922
last_value:
1915
1932
last_value:
1922
1908
last_value:
1932
1927
last_value:
1908
1904
last_value:
1927
1924
last_value:
1904
1925
last_value:
1924
1916
last_value:
1925
1917
last_value:
1916
1929
last_value:
1917
1912
last_value:
1929
1907


1935
1924
last_value:
1925
1916
last_value:
1924
1929
last_value:
1916
1927
last_value:
1929
1934
last_value:
1927
1912
last_value:
1934
1925
last_value:
1912
1915
last_value:
1925
1934
last_value:
1915
1916
last_value:
1934
1923
last_value:
1916
1920
last_value:
1923
1952
last_value:
1920
1913
last_value:
1952
1901
last_value:
1913
1946
last_value:
1901
1926
last_value:
1946
1922
last_value:
1926
1931
last_value:
1922
1942
last_value:
1931
1919
last_value:
1942
1908
last_value:
1919
1912
last_value:
1908
1928
last_value:
1912
1925
last_value:
1928
1937
last_value:
1925
1943
last_value:
1937
1921
last_value:
1943
1930
last_value:
1921
1940
last_value:
1930
1942
last_value:
1940
1920
last_value:
1942
1939
last_value:
1920
1941
last_value:
1939
1943
last_value:
1941
1924
last_value:
1943
1937
last_value:
1924
1941
last_value:
1937
1933
last_value:
1941
1925
last_value:
1933
1924
last_value:
1925
1941
last_value:
1924
1924
last_value:
1941
1944
last_value:
1924
1926
last_value:
1944
1920


1930
1957
last_value:
1935
1940
last_value:
1957
1948
last_value:
1940
1948
last_value:
1948
1932
last_value:
1948
1952
last_value:
1932
1940
last_value:
1952
1955
last_value:
1940
1952
last_value:
1955
1948
last_value:
1952
1953
last_value:
1948
1959
last_value:
1953
1956
last_value:
1959
1945
last_value:
1956
1948
last_value:
1945
1948
last_value:
1948
1935
last_value:
1948
1932
last_value:
1935
1933
last_value:
1932
1953
last_value:
1933
1934
last_value:
1953
1943
last_value:
1934
1948
last_value:
1943
1941
last_value:
1948
1947
last_value:
1941
1936
last_value:
1947
1947
last_value:
1936
1947
last_value:
1947
1957
last_value:
1947
1952
last_value:
1957
1949
last_value:
1952
1930
last_value:
1949
1922
last_value:
1930
1951
last_value:
1922
1942
last_value:
1951
1932
last_value:
1942
1959
last_value:
1932
1942
last_value:
1959
1931
last_value:
1942
1950
last_value:
1931
1934
last_value:
1950
1957
last_value:
1934
1939
last_value:
1957
1954
last_value:
1939
1945
last_value:
1954
1948


In [43]:
last_value

1963

In [45]:
legi_dataset.header

['last_name', 'first_name', 'birthday', 'gender', 'type', 'state', 'party']

[['Bassett', 'Richard', '1745-04-02', 'M', 'sen', 'DE', 'Anti-Administration', 1745], ['Bland', 'Theodorick', '1742-03-21', 'M', 'rep', 'VA', '', 1742], ['Burke', 'Aedanus', '1743-06-16', 'M', 'rep', 'SC', '', 1743], ['Carroll', 'Daniel', '1730-07-22', 'M', 'rep', 'MD', '', 1730], ['Clymer', 'George', '1739-03-16', 'M', 'rep', 'PA', '', 1739], ['Contee', 'Benjamin', '', 'M', 'rep', 'MD', '', <span style="color:Red">0</span>], ['Dalton', 'Tristram', '1738-05-28', 'M', 'sen', 'MA', 'Pro-Administration', 1738], ['Elmer', 'Jonathan', '1745-11-29', 'M', 'sen', 'NJ', 'Pro-Administration', 1745], ['Few', 'William', '1748-06-08', 'M', 'sen', 'GA', 'Anti-Administration', 1748], ['Floyd', 'William', '1734-12-17', 'M', 'rep', 'NY', '', 1734]]

In [42]:
print(legislators[:10])

[['Bassett', 'Richard', '1745-04-02', 'M', 'sen', 'DE', 'Anti-Administration', 1745], ['Bland', 'Theodorick', '1742-03-21', 'M', 'rep', 'VA', '', 1742], ['Burke', 'Aedanus', '1743-06-16', 'M', 'rep', 'SC', '', 1743], ['Carroll', 'Daniel', '1730-07-22', 'M', 'rep', 'MD', '', 1730], ['Clymer', 'George', '1739-03-16', 'M', 'rep', 'PA', '', 1739], ['Contee', 'Benjamin', '', 'M', 'rep', 'MD', '', 1739], ['Dalton', 'Tristram', '1738-05-28', 'M', 'sen', 'MA', 'Pro-Administration', 1738], ['Elmer', 'Jonathan', '1745-11-29', 'M', 'sen', 'NJ', 'Pro-Administration', 1745], ['Few', 'William', '1748-06-08', 'M', 'sen', 'GA', 'Anti-Administration', 1748], ['Floyd', 'William', '1734-12-17', 'M', 'rep', 'NY', '', 1734]]


In [46]:
for row in legislators:
    if row[6] == "":
        row[6] = "No Party"

In [47]:
print(legislators[:10])

[['Bassett', 'Richard', '1745-04-02', 'M', 'sen', 'DE', 'Anti-Administration', 1745], ['Bland', 'Theodorick', '1742-03-21', 'M', 'rep', 'VA', 'No Party', 1742], ['Burke', 'Aedanus', '1743-06-16', 'M', 'rep', 'SC', 'No Party', 1743], ['Carroll', 'Daniel', '1730-07-22', 'M', 'rep', 'MD', 'No Party', 1730], ['Clymer', 'George', '1739-03-16', 'M', 'rep', 'PA', 'No Party', 1739], ['Contee', 'Benjamin', '', 'M', 'rep', 'MD', 'No Party', 1739], ['Dalton', 'Tristram', '1738-05-28', 'M', 'sen', 'MA', 'Pro-Administration', 1738], ['Elmer', 'Jonathan', '1745-11-29', 'M', 'sen', 'NJ', 'Pro-Administration', 1745], ['Few', 'William', '1748-06-08', 'M', 'sen', 'GA', 'Anti-Administration', 1748], ['Floyd', 'William', '1734-12-17', 'M', 'rep', 'NY', 'No Party', 1734]]
